rclpy/src/rclpy/_rclpy.c

```diff
@@ -5436,6 +5435,23 @@ rclpy_service_info_get_received_timestamp(PyObject * Py_UNUSED(self), PyObject *
   return PyLong_FromLongLong(service_info->received_timestamp);
 }
 
+/// Retrieve the topic name from a rclpy_publisher_t
+static PyObject *
+rclpy_publisher_get_topic_name(PyObject * Py_UNUSED(self), PyObject * args)
+{
+    PyObject * pyentity;
+    if (!PyArg_ParseTuple(args, "O", &pyentity)) {
+        return NULL;
+    }
+
+    rclpy_publisher_t * publisher = rclpy_handle_get_pointer_from_capsule(
+        pyentity, "rclpy_publisher_t");
+    if (!publisher) {
+        return NULL;
+    }
+    return PyUnicode_FromString(rcl_publisher_get_topic_name(&publisher->publisher));
```

Comment:
One more thing to handle. It's possible `rcl_publisher_get_topic_name()` could fail, in which case the return value is `NULL` and an `rcl` error message is set. If this happens, I think we should raise an exception - and there's a custom exception for errors in `rcl` stored in the variable `RCLError` in this file.


Here's an example:

https://github.com/ros2/rclpy/blob/8f601c69ce9775ee3764c2bcfa2cd22cd9c31a1d/rclpy/src/rclpy/_rclpy.c#L184-L188

`PyErr_Format()` creates the exception of type `RCLError`, `rcl_get_error_string()` gets the error message set by `rcl_publisher_get_topic_name()`, and `rcl_reset_error()` clears the error message so it doesn't get in the way of the next RCL call. Mind adding that here?

Reply:
That makes sense, I'll add the handle here.